[Data Cleaning] Make commit_data_cleaning task more robust

biyeun opened the pull request:
## Technical Summary
This ensures the `commit_data_cleaning` task is more robust by
- adding logging to troubleshoot any issues
- catching partial errors, so that the entire task doesn't fail
- ensuring only one task can "claim" processing of the bulk cleaning session
- populating task_id and committed_on from within the task, rather than outside of it

## Feature Flag
`DATA_CLEANING_CASES`

## Safety Assurance

### Safety story
existing tests already test the functionality of `commit_data_cleaning`, these changes improve overall safety and error handling.

### Automated test coverage
yes, with room for improvements in a followup

### QA Plan
not yet (QA starting on overall feature this week, does not block PR)


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun committed:
remove session.save() as self.save() is already called at the end of update_result

biyeun committed:
completed_on should use timezone.now(), like committed_on

biyeun committed:
set task id and committed_on from within the task
and safeguard task from being picked up by multiple celery workers

biyeun committed:
properly iterate over chunked records
avoid O(N^2) behavior of large offset slicing and clean up overall logic

biyeun committed:
make purging UI details from session more resilient against concurrency issues

biyeun committed:
log when tasks are dropped to avoid duplication

biyeun committed:
log when tasks begin

biyeun committed:
log (unusual) case when no case updates are found
for troubleshooting purposes

biyeun committed:
ensure save only does a partial update of session to avoid clashes during task
note: we don't care about last_modified here, so fine to skip auto_now on that field

biyeun committed:
allow result to also track errors

biyeun committed:
small visual refactor

biyeun committed:
prune only un-errored records and related changes before marking session as complete

biyeun committed:
log completed session

biyeun committed:
catch errors in get_edited_case_properties

mjriley commented on corehq/apps/data_cleaning/tasks.py:
```diff
@@ -116,13 +116,27 @@ def _prune_records_and_complete_session(session, errored_doc_ids):
-        update = record.get_edited_case_properties(case)
+        try:
+            update = record.get_edited_case_properties(case)
+        except Exception as e:
```
can we add a comment justifying the catch-all exception here?

mjriley commented on corehq/apps/data_cleaning/tasks.py:
```diff
@@ -116,13 +116,27 @@ def _prune_records_and_complete_session(session, errored_doc_ids):
+        try:
+            update = record.get_edited_case_properties(case)
+        except Exception as e:
+            errored_doc_ids.append(record.doc_id)
```
might make sense to extract this into a utility method to ease readability

biyeun replied:
done https://github.com/dimagi/commcare-hq/pull/36358/commits/6c67945a96abc7bad7b14e259eea42a62fa67652

mjriley commented on corehq/apps/data_cleaning/tasks.py:
```diff
@@ -65,6 +61,15 @@ def commit_data_cleaning(self, bulk_edit_session_id):
     return form_ids
 
 
+@retry_on_integrity_error(max_retries=3, delay=0.1)
```
I'm not sure we need this retry decorator

biyeun replied:
removed this any other retries

mjriley reviewed (approved):
No blocking changes for now

biyeun committed:
improve readability around try/except with utility
also add todo about catching more explicit errors after some testing

biyeun committed:
improve readability around try/except
and add todo comment to catch more specific errors after some testing

biyeun committed:
remove acks_late=True and allow whole task to fail and not retry if something goes wrong

biyeun committed:
move deselect_all into purge ui utility and remove over-eager retry

biyeun committed:
remove over-eager retry

biyeun committed:
move claim session for task into a utility
- clarifies what we are doing at the beginning
- makes the over all core task code shorter and easier to read

biyeun committed:
prune completed records after each chunk submission
- retain the change history (for now)

biyeun committed:
move multi-line log to utility to reduce readability distractions

biyeun committed:
ordering isn't necessary, remove potential overhead

biyeun committed:
only prune records whose change_block didn't error

biyeun committed:
ensure that orphaned changes are also removed once records are pruned

biyeun committed:
fix tests

biyeun committed:
add required es adapter for test to make sure it's not flaky failing

mjriley commented on corehq/apps/data_cleaning/tasks.py:
```diff
@@ -38,7 +38,7 @@ def commit_data_cleaning(self, bulk_edit_session_id):
     session.update_result(0)
     count_cases = case_load_counter("bulk_case_cleaning", session.domain)
 
-    record_iter = session.records.order_by('pk').iterator()
```
Are we sure we don't need ordering here? Is it possible for the user to add or delete records for this session while this task is running?

biyeun replied:
I'm pretty sure it isn't necessary

> Is it possible for the user to add or delete records for this session while this task is running?
It isn't possible, the session will be locked down after it is committed

mjriley commented on corehq/apps/data_cleaning/tasks.py:
```diff
@@ -42,17 +44,17 @@ def commit_data_cleaning(self, bulk_edit_session_id):
+    record_iter = session.records.order_by('pk').iterator()
+    for record_batch in chunked(record_iter, CASEBLOCK_CHUNKSIZE):
+        blocks = _create_case_blocks(session, record_batch)
         xform = _submit_case_blocks(session, blocks)
```
I think I might be confused about the code here. We're iterating over all the records in a session, which likely represent multiple forms, right? And each `record_batch` also likely represents multiple forms? How can we be sure that `_submit_case_blocks(session, blocks)` only creates a single `xform` then?

biyeun replied:
This is a question for @orangejenny, but I believe the answer is that this is a bulk case update form and is already part of the case importer infrastructure

mjriley reviewed (approved):
Appreciate the readability changes. Still not sure about the bulk code representing an xform, but it seems that's in line with Jenny's understanding, so I'll sign off on this PR for now
